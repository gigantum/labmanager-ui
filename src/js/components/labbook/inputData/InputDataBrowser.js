// vendor
import React, { Component } from 'react';
import { createPaginationContainer, graphql } from 'react-relay';
// mutations
import FileBrowserWrapper from 'Components/labbook/fileBrowser/FileBrowserWrapper';
// store
import store from 'JS/redux/store';

class InputDataBrowser extends Component {
  constructor(props) {
  	super(props);

    const { owner, labbookName } = store.getState().routes;

    this.state = {
      show: false,
      message: '',
      files: [],
      moreLoading: false,
      owner,
      labbookName,
    };
  }

  /*
    loads more if branches are switched
  */
  componentDidUpdate() {
    this.props.loadStatus(this.state.moreLoading);
    if (!this.state.moreLoading && this.props.input.allFiles && this.props.input.allFiles.edges.length < 3 && this.props.input.allFiles.pageInfo.hasNextPage) {
      this._loadMore();
    }
  }

  /*
    handle state and addd listeners when component mounts
  */
  componentDidMount() {
    this.props.loadStatus(this.state.moreLoading);
    if (this.props.input.allFiles &&
      this.props.input.allFiles.pageInfo.hasNextPage) {
      this._loadMore();
    } else {
      this.setState({ moreLoading: false });
    }
  }

  /*
    @param
    triggers relay pagination function loadMore
    increments by 100
    logs callback
  */

  _loadMore() {
    // this.setState({'moreLoading': true});
    const self = this;
    this.props.relay.loadMore(
      100, // Fetch the next 100 feed items
      (response, error) => {
        if (error) {
          console.error(error);
        }

        if (self.props.input.allFiles &&
         self.props.input.allFiles.pageInfo.hasNextPage) {
          self._loadMore();
        } else {
          this.setState({ moreLoading: false });
        }
      },
    );
  }

  render() {
    if (this.props.input && this.props.input.allFiles) {
      let inputFiles = this.props.input.allFiles;
      if (this.props.input.allFiles.edges.length === 0) {
        inputFiles = {
          edges: [],
          pageInfo: this.props.input.allFiles.pageInfo,
        };
      }

      return (
        <FileBrowserWrapper
          ref="inputBrowser"
          section="input"
          files={inputFiles}
          selectedFiles={this.props.selectedFiles}
          clearSelectedFiles={this.props.clearSelectedFiles}
          connection="InputDataBrowser_allFiles"
          parentId={this.props.inputId}
          favoriteConnection="InputFavorites_favorites"
          owner={this.state.owner}
          isLocked={this.props.isLocked}
          {...this.props}
        />
      );
    }
    return (<div>No Files Found</div>);
  }
}


export default createPaginationContainer(
  InputDataBrowser,
  {

    input: graphql`
      fragment InputDataBrowser_input on LabbookSection{
        allFiles(after: $cursor, first: $first)@connection(key: "InputDataBrowser_allFiles", filters: []){
          edges{
            node{
              id
              isDir
              isFavorite
              modifiedAt
              key
              size
            }
            cursor
          }
          pageInfo{
            hasNextPage
            hasPreviousPage
            startCursor
            endCursor
          }
        }

      }`,
  },
  {
    direction: 'forward',
    getConnectionFromProps(props) {
      return props.input && props.input.allFiles;
    },
    getFragmentVariables(prevVars, totalCount) {
      return {
        ...prevVars,
        count: totalCount,
      };
    },
    getVariables(props, { count, cursor }, fragmentVariables) {
      const { owner, labbookName } = store.getState().routes;

      return {
        first: count,
        cursor,
        owner,
        name: labbookName,
      };
    },
    query: graphql`
      query InputDataBrowserPaginationQuery(
        $first: Int
        $cursor: String
        $owner: String!
        $name: String!
      ) {
        labbook(name: $name, owner: $owner){
          input{
            # You could reference the fragment defined previously.
            ...InputDataBrowser_input
          }
        }
      }
    `,
  },
);
