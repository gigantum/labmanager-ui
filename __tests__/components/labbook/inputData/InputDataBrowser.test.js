
import React, {Component} from 'react'
import renderer from 'react-test-renderer';
import {mount} from 'enzyme'
import InputDataBrowser from 'Components/labbook/inputData/InputDataBrowser';

import { DragDropContext } from 'react-dnd'
import HTML5Backend from 'react-dnd-html5-backend'

import json from './__relaydata__/InputDataBrowser.json'

import relayTestingUtils from 'relay-testing-utils'
      const loadStatus = ()=>{

      }


      const clearSelectedFiles = ()=>{

      }

      const backend = (manager: Object) => {
          const backend = HTML5Backend(manager),
              orgTopDropCapture = backend.handleTopDropCapture;

          backend.handleTopDropCapture = (e) => {

              if(backend.currentNativeSource){
                orgTopDropCapture.call(backend, e);

               //backend.currentNativeSource.item.dirContent = getFilesFromDragEvent(e, {recursive: true}); //returns a promise
              }
          };

          return backend;
      }

      const fixtures = {
        labbook: json.data.labbook,
        labbookId: json.data.labbook.id,
        isLocked: false,
        selectedFiles: [],
        clearSelectedFiles,
        inputId: json.data.labbook.input.id,
        input: json.data.labbook.input,
        loadStatus
      }

      class InputDataCompInstance extends Component{
        render(){
          return (relayTestingUtils.relayWrap(<InputDataBrowser {...fixtures}/>, {}, json.data.labbook))
        }
      }
      const InputDataBrowserComponent = DragDropContext(backend)(InputDataCompInstance)

      describe('Test CodeBrowser', () => {

        it('snapshot renders', () => {
          const wrapper = renderer.create(
             <InputDataBrowserComponent />
          );

          const tree = wrapper.toJSON()

          expect(tree).toMatchSnapshot()
        })


      })
