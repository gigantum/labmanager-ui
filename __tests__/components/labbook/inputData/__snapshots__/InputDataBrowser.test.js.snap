// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test CodeBrowser snapshot renders 1`] = `
<div
  className="Code flex flex-row justify-center"
  id="code"
>
  <div
    className="rendered-react-keyed-file-browser"
  >
    <div
      className="rendered-file-browser"
    >
      <div
        className="action-bar"
      >
        <input
          onChange={[Function]}
          placeholder="Filter files"
          type="search"
          value=""
        />
        <ul
          className="item-actions"
        >
          <li>
            <a
              className="btn btn-primary btn-sm"
              href="#"
              onClick={[Function]}
              role="button"
            >
              <i
                aria-hidden="true"
                className="fa fa-folder-o"
              />
              Â Add Folder
            </a>
          </li>
        </ul>
      </div>
      <div
        className="files"
        id="InputDataBrowser_allFiles"
      >
        <table
          cellPadding="0"
          cellSpacing="0"
        >
          <thead>
            <tr
              className="folder"
            >
              <th>
                File
              </th>
              <th />
              <th
                className="size"
              >
                Size
              </th>
              <th
                className="modified"
              >
                Last Modified
              </th>
              <th />
            </tr>
          </thead>
          <tbody>
            <tr
              className="file modified"
              onClick={[Function]}
              onDoubleClick={[Function]}
            >
              <td
                className="name"
              >
                <div
                  style={
                    Object {
                      "paddingLeft": "0px",
                    }
                  }
                >
                  <div>
                    <a
                      download="download"
                      href="#"
                      onClick={[Function]}
                    >
                      <span
                        className="FileBrowser__icon graph-icon"
                      />
                      nyc_crime_2.csv
                    </a>
                  </div>
                </div>
              </td>
              <td
                width="30"
              >
                <div
                  className="Favorite__star--empty"
                  onClick={[Function]}
                />
              </td>
              <td
                className="size"
              >
                206.55 MB
              </td>
              <td
                className="modified"
              >
                2 months ago
              </td>
            </tr>
            <tr
              className="file modified"
              onClick={[Function]}
              onDoubleClick={[Function]}
            >
              <td
                className="name"
              >
                <div
                  style={
                    Object {
                      "paddingLeft": "0px",
                    }
                  }
                >
                  <div>
                    <a
                      download="download"
                      href="#"
                      onClick={[Function]}
                    >
                      <span
                        className="FileBrowser__icon graph-icon"
                      />
                      nyc_taxi.csv
                    </a>
                  </div>
                </div>
              </td>
              <td
                width="30"
              >
                <div
                  className="Favorite__star--empty"
                  onClick={[Function]}
                />
              </td>
              <td
                className="size"
              >
                1440.63 MB
              </td>
              <td
                className="modified"
              >
                2 months ago
              </td>
            </tr>
            <tr
              className="file modified"
              onClick={[Function]}
              onDoubleClick={[Function]}
            >
              <td
                className="name"
              >
                <div
                  style={
                    Object {
                      "paddingLeft": "0px",
                    }
                  }
                >
                  <div>
                    <a
                      download="download"
                      href="#"
                      onClick={[Function]}
                    >
                      <span
                        className="FileBrowser__icon jupyter-icon"
                      />
                      nyc_taxi.ipynb
                    </a>
                  </div>
                </div>
              </td>
              <td
                width="30"
              >
                <div
                  className="Favorite__star--empty"
                  onClick={[Function]}
                />
              </td>
              <td
                className="size"
              >
                14.54 MB
              </td>
              <td
                className="modified"
              >
                2 months ago
              </td>
            </tr>
            <tr
              className="file modified"
              onClick={[Function]}
              onDoubleClick={[Function]}
            >
              <td
                className="name"
              >
                <div
                  style={
                    Object {
                      "paddingLeft": "0px",
                    }
                  }
                >
                  <div>
                    <a
                      download="download"
                      href="#"
                      onClick={[Function]}
                    >
                      <span
                        className="FileBrowser__icon zip-icon"
                      />
                      nyc_taxi.zip
                    </a>
                  </div>
                </div>
              </td>
              <td
                width="30"
              >
                <div
                  className="Favorite__star--empty"
                  onClick={[Function]}
                />
              </td>
              <td
                className="size"
              >
                412.27 MB
              </td>
              <td
                className="modified"
              >
                2 months ago
              </td>
            </tr>
            <tr
              className="file modified"
              onClick={[Function]}
              onDoubleClick={[Function]}
            >
              <td
                className="name"
              >
                <div
                  style={
                    Object {
                      "paddingLeft": "0px",
                    }
                  }
                >
                  <div>
                    <a
                      download="download"
                      href="#"
                      onClick={[Function]}
                    >
                      <span
                        className="FileBrowser__icon zip-icon"
                      />
                      v2.zip
                    </a>
                  </div>
                </div>
              </td>
              <td
                width="30"
              >
                <div
                  className="Favorite__star--empty"
                  onClick={[Function]}
                />
              </td>
              <td
                className="size"
              >
                4.79 MB
              </td>
              <td
                className="modified"
              >
                2 months ago
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
  <div
    className="DetailPanel hidden"
  >
    <div
      className="DetailPanel--close"
      onClick={[Function]}
    >
      X
    </div>
    <p />
    <p />
  </div>
</div>
`;
